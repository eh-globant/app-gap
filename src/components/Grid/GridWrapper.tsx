import { useState, useEffect, useCallback } from "react";
import Grid from "./Grid";
import { Product } from "../../types/product";
import useInfiniteScroll from "../../hooks/useInfiniteScroll";
import allProducts from "../../data/products";
import styles from "./Grid.module.scss";

interface GridProps {
  productsPerPage?: number;
}

export default function GridWrapper({ productsPerPage = 8 }: GridProps) {
  const [visibleProducts, setVisibleProducts] = useState<Product[]>([]);
  const [page, setPage] = useState(0);
  const [isLoading, setIsLoading] = useState(false);

  // Load more products
  const loadMoreProducts = useCallback(async () => {
    if (isLoading || visibleProducts.length >= allProducts.length) return;

    setIsLoading(true);

    // Simulating network delay
    await new Promise((resolve) => setTimeout(resolve, 1000));

    // Calculate indexes
    const startIndex = page * productsPerPage;
    const endIndex = (page + 1) * productsPerPage;
    const nextProducts = allProducts.slice(startIndex, endIndex);

    // Add new products
    setVisibleProducts((prev) => [...prev, ...nextProducts]);
    setPage((prevPage) => prevPage + 1);
    setIsLoading(false);
  }, [page, isLoading, productsPerPage]);

  // Initial load
  useEffect(() => {
    // Load the first products at start (8 by default)
    const initialProducts = allProducts.slice(0, productsPerPage);
    setVisibleProducts(initialProducts);
    setPage(1);
  }, [productsPerPage]);

  const targetRef = useInfiniteScroll(
    loadMoreProducts,
    visibleProducts.length < allProducts.length,
    isLoading
  );

  return (
    <>
      <div className={styles["grid__wrapper"]} data-testid="grid-wrapper">
        <Grid products={visibleProducts} />
      </div>
      <div ref={targetRef} className={styles["grid__scroll-trigger"]}>
        {isLoading && (
          <div className={styles["grid__loading-spinner"]}>Loading...</div>
        )}
      </div>
    </>
  );
}
