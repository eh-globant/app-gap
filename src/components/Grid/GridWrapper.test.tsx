import { render, waitFor, act } from "@testing-library/react";
import GridWrapper from "./GridWrapper";
import "@testing-library/jest-dom";

// Type for mock of IntersectionObserver
interface MockIntersectionObserver {
  new (
    callback: IntersectionObserverCallback,
    options?: IntersectionObserverInit
  ): {
    observe: jest.Mock;
    unobserve: jest.Mock;
    disconnect: jest.Mock;
  };
}

// Mock products
jest.mock("../../data/products", () => ({
  __esModule: true,
  default: Array.from({ length: 32 }, (_, i) => ({
    id: `prod-${i + 1}`,
    name: `Product ${i + 1}`,
    image: `image-${i + 1}.jpg`,
    description: `Description ${i + 1}`,
    cost: 100 * (i + 1),
    discount: 80 * (i + 1),
  })),
}));

describe("GridWrapper Component", () => {
  let mockObserve: jest.Mock;
  let mockUnobserve: jest.Mock;
  let mockDisconnect: jest.Mock;
  let observerCallback: IntersectionObserverCallback;
  let observerInstance: any;

  beforeEach(() => {
    // Reset mocks
    jest.clearAllMocks();

    // Initialize mocks
    mockObserve = jest.fn();
    mockUnobserve = jest.fn();
    mockDisconnect = jest.fn();

    // Configurare IntersectionObserver mock
    class MockIntersectionObserver {
      constructor(callback: IntersectionObserverCallback) {
        observerCallback = callback;
        observerInstance = this;
      }
      observe = mockObserve;
      unobserve = mockUnobserve;
      disconnect = mockDisconnect;
    }

    // Asign mock
    (window as any).IntersectionObserver = MockIntersectionObserver;
  });

  // Helper to simulate scroll
  const simulateScroll = async () => {
    await act(async () => {
      const entry: IntersectionObserverEntry = {
        isIntersecting: true,
        intersectionRatio: 1,
        boundingClientRect: {} as DOMRectReadOnly,
        intersectionRect: {} as DOMRectReadOnly,
        rootBounds: null,
        target: document.createElement("div"),
        time: 0,
      };
      observerCallback([entry], observerInstance);
    });
  };

  it("renders without crashing", () => {
    const { getByTestId } = render(<GridWrapper />);
    expect(getByTestId("grid-wrapper")).toBeInTheDocument();
  });

  it("loads initial 8 products", async () => {
    const { getAllByRole } = render(<GridWrapper />);

    await waitFor(() => {
      expect(getAllByRole("img")).toHaveLength(8);
    });
  });

  it("shows loading indicator when fetching more products", async () => {
    const { getByText, queryByText } = render(<GridWrapper />);

    await simulateScroll();
    expect(getByText("Loading...")).toBeInTheDocument();

    await waitFor(
      () => {
        expect(queryByText("Loading...")).not.toBeInTheDocument();
      },
      { timeout: 2000 }
    );
  });

  it("loads additional 8 products when scrolled", async () => {
    const { getAllByRole } = render(<GridWrapper />);

    // Verify initial load
    await waitFor(() => {
      expect(getAllByRole("img")).toHaveLength(8);
    });

    // Simulate scroll
    await simulateScroll();

    // Verify additional load
    await waitFor(
      () => {
        expect(getAllByRole("img")).toHaveLength(16);
      },
      { timeout: 2000 }
    );
  });

  it("matches snapshot", async () => {
    const { asFragment, getAllByRole } = render(<GridWrapper />);

    await waitFor(() => {
      expect(getAllByRole("img")).toHaveLength(8);
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
