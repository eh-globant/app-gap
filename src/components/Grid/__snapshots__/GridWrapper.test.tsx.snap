// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridWrapper Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid__wrapper"
    data-testid="grid-wrapper"
  >
    <div
      class="grid"
    >
      <div
        class="card__container"
      >
        <div
          class="card__image-container"
        >
          <img
            alt="Product 1"
            class="card__image"
            src="image-1.jpg"
          />
        </div>
        <div
          class="card__info-container"
        >
          <p
            class="card__info-description"
          >
            Description 1
          </p>
          <p
            class="card__info-cost"
          >
            $100.00
          </p>
          <p
            class="card__info-discount"
          >
            $80.00
          </p>
        </div>
      </div>
      <div
        class="card__container"
      >
        <div
          class="card__image-container"
        >
          <img
            alt="Product 2"
            class="card__image"
            src="image-2.jpg"
          />
        </div>
        <div
          class="card__info-container"
        >
          <p
            class="card__info-description"
          >
            Description 2
          </p>
          <p
            class="card__info-cost"
          >
            $200.00
          </p>
          <p
            class="card__info-discount"
          >
            $160.00
          </p>
        </div>
      </div>
      <div
        class="card__container"
      >
        <div
          class="card__image-container"
        >
          <img
            alt="Product 3"
            class="card__image"
            src="image-3.jpg"
          />
        </div>
        <div
          class="card__info-container"
        >
          <p
            class="card__info-description"
          >
            Description 3
          </p>
          <p
            class="card__info-cost"
          >
            $300.00
          </p>
          <p
            class="card__info-discount"
          >
            $240.00
          </p>
        </div>
      </div>
      <div
        class="card__container"
      >
        <div
          class="card__image-container"
        >
          <img
            alt="Product 4"
            class="card__image"
            src="image-4.jpg"
          />
        </div>
        <div
          class="card__info-container"
        >
          <p
            class="card__info-description"
          >
            Description 4
          </p>
          <p
            class="card__info-cost"
          >
            $400.00
          </p>
          <p
            class="card__info-discount"
          >
            $320.00
          </p>
        </div>
      </div>
      <div
        class="card__container"
      >
        <div
          class="card__image-container"
        >
          <img
            alt="Product 5"
            class="card__image"
            src="image-5.jpg"
          />
        </div>
        <div
          class="card__info-container"
        >
          <p
            class="card__info-description"
          >
            Description 5
          </p>
          <p
            class="card__info-cost"
          >
            $500.00
          </p>
          <p
            class="card__info-discount"
          >
            $400.00
          </p>
        </div>
      </div>
      <div
        class="card__container"
      >
        <div
          class="card__image-container"
        >
          <img
            alt="Product 6"
            class="card__image"
            src="image-6.jpg"
          />
        </div>
        <div
          class="card__info-container"
        >
          <p
            class="card__info-description"
          >
            Description 6
          </p>
          <p
            class="card__info-cost"
          >
            $600.00
          </p>
          <p
            class="card__info-discount"
          >
            $480.00
          </p>
        </div>
      </div>
      <div
        class="card__container"
      >
        <div
          class="card__image-container"
        >
          <img
            alt="Product 7"
            class="card__image"
            src="image-7.jpg"
          />
        </div>
        <div
          class="card__info-container"
        >
          <p
            class="card__info-description"
          >
            Description 7
          </p>
          <p
            class="card__info-cost"
          >
            $700.00
          </p>
          <p
            class="card__info-discount"
          >
            $560.00
          </p>
        </div>
      </div>
      <div
        class="card__container"
      >
        <div
          class="card__image-container"
        >
          <img
            alt="Product 8"
            class="card__image"
            src="image-8.jpg"
          />
        </div>
        <div
          class="card__info-container"
        >
          <p
            class="card__info-description"
          >
            Description 8
          </p>
          <p
            class="card__info-cost"
          >
            $800.00
          </p>
          <p
            class="card__info-discount"
          >
            $640.00
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="grid__scroll-trigger"
  />
</DocumentFragment>
`;
